contracts/strategies/euler/IdleEulerStakingStrategy.sol

```diff
@@ -0,0 +1,227 @@
+
+        uint256 underlyingsInEuler = eToken.balanceOfUnderlying(address(this));
+        if (_amountToWithdraw > underlyingsInEuler) {
+            _amountToWithdraw = underlyingsInEuler;
+        }
+
+        // Withdraw from Euler
+        uint256 underlyingBalanceBefore = _underlyingToken.balanceOf(address(this));
+        _eToken.withdraw(SUB_ACCOUNT_ID, _amountToWithdraw);
+        amountWithdrawn = _underlyingToken.balanceOf(address(this)) - underlyingBalanceBefore;
+        // Send tokens to the destination
+        _underlyingToken.safeTransfer(_destination, amountWithdrawn);
+    }
+
+    /// @return rewards rewards[0] : rewards redeemed
+    function _redeemRewards(bytes calldata) internal override returns (uint256[] memory rewards) {
+        // Get rewards from StakingRewards contract
+        stakingRewards.getReward();
+        // transfer rewards to the IdleCDO contract
+        rewards = new uint256[](1);
+        rewards[0] = EUL.balanceOf(address(this));
+        EUL.safeTransfer(idleCDO, rewards[0]);
+    }
+
+    // ###################
+    // Views
+    // ###################
+
+    /// @return net price in underlyings of 1 strategyToken
+    function price() public view override returns (uint256) {
+        uint256 eTokenDecimals = eToken.decimals();
+        // return price of 1 eToken in underlying
+        return eToken.convertBalanceToUnderlying(10**eTokenDecimals);
+    }
+
+    // TODO add staking apr 
+
+    /// @dev Returns supply apr for providing liquidity minus reserveFee
+    /// @return apr net apr (fees should already be excluded)
+    function getApr() external view override returns (uint256 apr) {
+        // Use the markets module:
+        IMarkets markets = IMarkets(EULER_MARKETS);
+        IDToken dToken = IDToken(markets.underlyingToDToken(token));
+        uint256 borrowSPY = uint256(int256(markets.interestRate(token)));
+        uint256 totalBorrows = dToken.totalSupply();
+        uint256 totalBalancesUnderlying = eToken.totalSupplyUnderlying();
+        uint32 reserveFee = markets.reserveFee(token);
+        // (borrowAPY, supplyAPY)
+        (, apr) = IEulerGeneralView(EULER_GENERAL_VIEW).computeAPYs(
+            borrowSPY,
+            totalBorrows,
+            totalBalancesUnderlying,
+            reserveFee
+        );
+        // apr is eg 0.024300334 * 1e27 for 2.43% apr
+        // while the method needs to return the value in the format 2.43 * 1e18
+        // so we do apr / 1e9 * 100 -> apr / 1e7
+        apr = apr / 1e7;
+    }
+
+    /// @return tokens array of reward token addresses
+    function getRewardTokens() external pure override returns (address[] memory tokens) {
+        tokens = new address[](1);
+        tokens[0] = address(EUL);
+    }
+
+    // ###################
+    // Protected
+    // ###################
+
+    ///@notice Claim rewards and withdraw all from StakingRewards contract
+    function exitStaking() external onlyOwner {
```

Comment:
in my intention this method is used for emergency.

Reply:
fair enough